sentry/src/main/java/io/sentry/MetricsAggregator.java

```diff
@@ -153,49 +160,49 @@ private void add(
       final double value,
       @Nullable MeasurementUnit unit,
       final @Nullable Map<String, String> tags,
-      @Nullable Long timestampMs,
+      @NotNull Long timestampMs,
       final int stackLevel) {
 
     if (isClosed) {
       return;
     }
 
-    if (timestampMs == null) {
-      timestampMs = nowMillis();
-    }
-
-    final @NotNull Metric metric;
-    switch (type) {
-      case Counter:
-        metric = new CounterMetric(key, value, unit, tags, timestampMs);
-        break;
-      case Gauge:
-        metric = new GaugeMetric(key, value, unit, tags, timestampMs);
-        break;
-      case Distribution:
-        metric = new DistributionMetric(key, value, unit, tags, timestampMs);
-        break;
-      case Set:
-        metric = new SetMetric(key, unit, tags, timestampMs);
-        //noinspection unchecked
-        metric.add((int) value);
-        break;
-      default:
-        throw new IllegalArgumentException("Unknown MetricType: " + type.name());
-    }
-
     final long timeBucketKey = MetricsHelper.getTimeBucketKey(timestampMs);
     final @NotNull Map<String, Metric> timeBucket = getOrAddTimeBucket(timeBucketKey);
 
     final @NotNull String metricKey = MetricsHelper.getMetricBucketKey(type, key, unit, tags);
 
-    // TODO check if we can synchronize only the metric itself
+    // TODO ideally we can synchronize only the metric itself
     synchronized (timeBucket) {
       @Nullable Metric existingMetric = timeBucket.get(metricKey);
       if (existingMetric != null) {
+        final int oldWeight = existingMetric.getWeight();
         existingMetric.add(value);
+        final int newWeight = existingMetric.getWeight();
+        totalBucketsWeight.addAndGet(newWeight - oldWeight);
       } else {
+        final @NotNull Metric metric;
+        switch (type) {
+          case Counter:
+            metric = new CounterMetric(key, value, unit, tags, timestampMs);
+            break;
+          case Gauge:
+            metric = new GaugeMetric(key, value, unit, tags, timestampMs);
+            break;
+          case Distribution:
+            metric = new DistributionMetric(key, value, unit, tags, timestampMs);
+            break;
+          case Set:
+            metric = new SetMetric(key, unit, tags, timestampMs);
+            // sets API is either ints or strings cr32 encoded into ints
+            // noinspection unchecked
+            metric.add((int) value);
+            break;
+          default:
+            throw new IllegalArgumentException("Unknown MetricType: " + type.name());
+        }
         timeBucket.put(metricKey, metric);
+        totalBucketsWeight.addAndGet(metric.getWeight());
       }
     }
 
```

Comment:
you should check here for the total weight, and in case flush everything `ahead of scheduled time`
Right now you are still waiting for the scheduled task to check the weight and flush it